### ISSUE TITLE
`[missing_const_for_fn]` is triggered when a function uses a `Drop` parameter

### ISSUE THREAD

**Issue by botahamec** [rust-lang/rust-clippy#9271]
### Summary

The lint says that the function below could be made into a `const fn`. Following this suggestion results in a compiler error, which points at x and says, "destructors cannot be evaluated at compile-time" 

### Lint Name

``[`missing_const_for_fn`]``

### Reproducer

I tried this code:

```rust
#![allow(clippy::all)]
#![allow(dead_code)]
#![warn(clippy::missing_const_for_fn)]

fn foo(x: Box<usize>) -> usize {
    *x + 5
}
```

I saw this happen:

```
    Checking playground v0.0.1 (/playground)
warning: this could be a `const fn`
 --> src/lib.rs:5:1
  |
5 | / fn foo(x: Box<usize>) -> usize {
6 | |     *x + 5
7 | | }
  | |_^
  |
note: the lint level is defined here
 --> src/lib.rs:3:9
  |
3 | #![warn(clippy::missing_const_for_fn)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn

warning: `playground` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.36s
```

I expected to see this happen: No output

### Version

```text
rustc 1.64.0-nightly (f8588549c 2022-07-18)
PS C:\Users\epice\Projects\lunars> rustc -Vv      
rustc 1.64.0-nightly (f8588549c 2022-07-18)
binary: rustc
commit-hash: f8588549c3c3d45c32b404210cada01e2a45def3
commit-date: 2022-07-18
host: x86_64-pc-windows-msvc
release: 1.64.0-nightly
LLVM version: 14.0.6
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by satvikpendem** [comment]
I get this often for some basic `String` conversion methods I have, so I have to do this each time:

```rust
#[allow(clippy::missing_const_for_fn)]
pub fn into_inner(self) -> String {
    self.0
}
```

**Comment by FreezyLemon** [comment]
I am not 100% sure if this is the same bug, but I think so. If you add this code to `tests/ui/missing_const_for_fn/cant_be_const.rs`:

```rust
mod destructor_runs_drop {
    struct D;
    struct S {
        inner: D,
    }

    enum E {
        Value(D),
    }

    impl Drop for D {
        fn drop(&mut self) {}
    }

    impl S {
        fn d(self) -> D {
            self.inner
        }
    }

    impl E {
        fn d(self) -> D {
            let Self::Value(d) = self;
            d
        }
    }

    fn struct_to_d(s: S) -> D {
        s.inner
    }

    fn tuple_to_d(t: (D, usize)) -> D {
        t.0
    }
}
```

You'll see that clippy thinks these functions can be `const fn`, even though they can't be. I think this is due to `S`, `E`, and the tuple not being `Drop` (making clippy think this can be `const fn`) but still needing to run `Drop` in their destructor (which causes the compiler to not allow `const fn`).

**Comment by JarredAllen** [comment]
Are there any plans to fix this false positive? Or are we just waiting for https://github.com/rust-lang/rust/issues/73255 to let these functions be `const`? I've just hit this issue with something similar to @satvikpendem's example.

**Comment by Centri3** [comment]
[It already has been.](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=52fdeb7bf5cb4c846b5b5f82c289a69d) #10891

**Comment by JarredAllen** [comment]
Neat! Looking forward to when it hits stable!
